ALSpringBackend/src/main/java/com/ardaslegends/data/service/Pathfinder.java

```diff
@@ -0,0 +1,156 @@
+package com.ardaslegends.data.service;
+
+import com.ardaslegends.data.domain.*;
+import com.ardaslegends.data.repository.RegionRepository;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+
+@RequiredArgsConstructor
+
+@Service
+public class Pathfinder {
+  private final RegionRepository _regionRepository;
+
+  /**
+   * Find the shortest path
+   * Return an object which contains the path and sum of weights
+   */
+
+  public Path findShortestWay(
+    String startRegionID,
+    String endRegionID,
+    Player player,
+    boolean isLeaderMove
+  ){
+    Optional<Region> startRegionOpt = this._regionRepository.findById(startRegionID);
+    Optional<Region> endRegionOpt = this._regionRepository.findById(endRegionID);
+    if(startRegionOpt.isEmpty() || endRegionOpt.isEmpty()){
+      return null;
+    }
+    Region startRegion = startRegionOpt.get();
+    Region endRegion = endRegionOpt.get();
+    //smallest weights between startRegion and all the other nodes
+    Map<Region, Integer> smallestWeights = new HashMap<>();
+    //for convenience, mark distance from startRegion to itself as 0
+    smallestWeights.put(startRegion, 0);
+
+    //implicit graph of all nodes and previous node in ideal paths
+    Map<Region, Region> prevNodes = new HashMap<>();
+
+    //use queue for breadth first search
+    //for convenience, we'll use an array, but linked list would be preferred
+    ArrayList<Region> nodesToVisitQueue= new ArrayList<>();
+
+    //record visited nodes with a set. The string is the toString of a HexID
+    Set<String> visitedNodes = new HashSet<>();
+    visitedNodes.add(startRegion.getId());
+
+    Region currentNode = startRegion;
+    //loop through nodes
+    while (true) {

```

Comment:
I think it's better to just do `while (nodesToVisitQueue.size() > 0)` here since you check that at the end of the loop anyways

Reply:
Thing is at the beginning the size is 0, and the first add is done in the while loop.

Reply:
I personally would put all the regions into the `nodesToVisitQueue` before you even start the loop. You then loop through all the regions in `nodesToVisitQueue` and calculate the stuff

Reply:
I don't have the code in front of me, but it's a broad search (if it's called like that), starting from initial node. So idk how it would behave if we just move randomly between nodes without any link. It's essentially pre-marking some nodes that we encountered in our current progress.

Reply:
ah yes, true
I don't know if there is another good way of fixing this then. I mean it doesn't really matter, not an impactful change, so we can leave it as it is now